src/audio/OpenALAudioEngine.cpp

```diff
@@ -0,0 +1,491 @@
+                    return State::Paused;
+                default:
+                    return State::Stopped;
+            }
+        };
+
+        void play() override
+        {
+            getSource();
+            alSourcePlay(m_source);
+        }
+
+        void pause() override
+        {
+            if (!m_hasSource) return;
+            getSource();
+            alSourcePause(m_source);
+        }
+
+        void stop() override
+        {
+            if (!m_hasSource) return;
+            getSource();
+            alSourceStop(m_source);
+        }
+
+    private:
+        ALuint m_buffer = 0;
+        bool m_hasBuffer;
+
+    protected:
+        ALuint m_source = 0;
+        bool m_hasSource = false; 
+        OpenALAudioEngine* m_engine;
+
+        virtual void getSource()
+        {
+            if (m_hasSource) return;
+            m_source = m_engine->getFreeSource();
+            m_hasSource = true;
+            alSourcei(m_source, AL_BUFFER, m_buffer);
+        }
+    };
+
+    class StreamingOpenALSound final
+        : public OpenALSound
+    {
+        /**
+         * @brief Number of buffers to use while streaming
+         * 
+         */
+        static constexpr std::size_t BUFFER_NUM = 3;
+
+        /**
+         * @brief Number of samples for each buffer
+         * 
+         * Longer buffers mean lower processing power required,
+         * but higher latency and increased memory usage
+         */
+        static constexpr std::size_t BUFFER_SIZE = 1024;
+    public:
+        StreamingOpenALSound(OpenALAudioEngine* engine,
+                            SoundStream stream, Format fmt, std::size_t sampleRate)
+            : OpenALSound(engine)
+            , m_stream(stream)
+            , m_format(fmt == Format::Mono ? AL_FORMAT_MONO16 : AL_FORMAT_STEREO16)
+            , m_sampleRate(sampleRate)
+            , m_stop(false)
+        {
+            alGenBuffers(BUFFER_NUM, m_buffers.data());
+            ALenum error = alGetError();
+            if (error != AL_NO_ERROR) throw OpenALException(error);
+
+            play();
+
+            m_thread = std::thread(&StreamingOpenALSound::renderThread, this);
+        }
+
+        ~StreamingOpenALSound() override
+        {
+            m_stop = true;
+            m_thread.join();
+            if (m_hasSource)
+            {
+                // Stop source and detach from buffer
+                alSourceStop(m_source);
+                ALint queuedBuffers;
+                alSourceUnqueueBuffers(m_source, BUFFER_NUM, m_buffers.data());
+                alSourcei(m_source, AL_BUFFER, 0);
+                m_engine->m_freeSources.push_back(m_source);
+            }
+            
+            alDeleteBuffers(BUFFER_NUM, m_buffers.data());
+        }
+
+    private:
+        SoundStream m_stream;
+        ALenum m_format;
+        std::size_t m_sampleRate;
+        std::atomic_bool m_stop;
+        std::array<ALuint, BUFFER_NUM> m_buffers{};
+        std::thread m_thread;
```

Comment:
Indeed it is a bit wasteful to have a separate thread for each streaming sound, it would have be nice to have a single thread that manages all streaming sounds at once, but it complicates the design as you said.

However, there should be very few streaming sounds going on generally (one for music, one for video sound playback when it will get implemented), so I don't think it is too bad of a perf hit.